cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
+	kube_api "k8s.io/kubernetes/pkg/api"
+)
+
+// AwsCloudProvider implements CloudProvider interface.
+type AwsCloudProvider struct {
+	awsManager *AwsManager
+	asgs       []*Asg
+}
+
+// BuildAwsCloudProvider builds CloudProvider implementation for AWS.
+func BuildAwsCloudProvider(awsManager *AwsManager, specs []string) (*AwsCloudProvider, error) {
+	aws := &AwsCloudProvider{
+		awsManager: awsManager,
+		asgs:       make([]*Asg, 0),
+	}
+	for _, spec := range specs {
+		if err := aws.addNodeGroup(spec); err != nil {
+			return nil, err
+		}
+	}
+	return aws, nil
+}
+
+// addNodeGroup adds node group defined in string spec. Format:
+// minNodes:maxNodes:asgName
+func (aws *AwsCloudProvider) addNodeGroup(spec string) error {
+	asg, err := buildAsg(spec, aws.awsManager)
+	if err != nil {
+		return err
+	}
+	aws.asgs = append(aws.asgs, asg)
+	aws.awsManager.RegisterAsg(asg)
+	return nil
+}
+
+// Name returns name of the cloud provider.
+func (aws *AwsCloudProvider) Name() string {
+	return "aws"
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (aws *AwsCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	result := make([]cloudprovider.NodeGroup, 0, len(aws.asgs))
+	for _, asg := range aws.asgs {
+		result = append(result, asg)
+	}
+	return result
+}
+
+// NodeGroupForNode returns the node group for the given node.
+func (aws *AwsCloudProvider) NodeGroupForNode(node *kube_api.Node) (cloudprovider.NodeGroup, error) {
+	ref, err := AwsRefFromProviderId(node.Spec.ProviderID)
+	if err != nil {
+		return nil, err
+	}
+	asg, err := aws.awsManager.GetAsgForInstance(ref)
+	return asg, err
+}
+
+// AwsRef contains a reference to some entity in AWS/GKE world.
+type AwsRef struct {
+	Zone string
```

Comment:
What does a zone represent?  If it is an availability zone, the autoscaling groups span availability zones, so zone is not part of the group's identity.